### PR TITLE
chore(db): bundle builder

### PR DESCRIPTION
**Pull request by emhane**
Merges changes from https://github.com/paradigmxyz/reth/pull/8264 and https://github.com/paradigmxyz/reth/pull/8861

### CODE REVIEW

File: crates/storage/provider/src/providers/database/provider.rs [hunk 1]
```diff
@@ -643,39 +649,59 @@ impl<TX: DbTxMut + DbTx> DatabaseProvider<TX> {
...
                 }
             }
-            // insert old info into reverts.
-            reverts.entry(block_number).or_default().entry(address).or_default().0 = Some(old_info);
+
+            // Add the old account info to the bundle builder if it exists.
+            if let Some(old_info) = old_info {
+                bundle_builder =
+                    bundle_builder.state_original_account_info(address, old_info.into());
+            }
```
Comment (emhane) [inline crates/storage/provider/src/providers/database/provider.rs:671]: review from https://github.com/paradigmxyz/reth/pull/8264/files#r1646535481 @rkrasiuk 

> previous behavior: if the entry is occuppied, overwrite the original account state even if old_info is None
current behavior: only overwrite the original account state if old_info is Some. 
if old_info is None, original info must still be overwritten

the methods `BundleBuilder::state_original_account_info` and `BundleBuilder::state_present_account_info` don't take an `Option<AccountInfo` as param. so I'm wondering:

- if entry is occupied but the original info is `None`, should we pass the `new_info` to `BundleBuilder::state_original_account_info`?
- there is another method `BundleBuilder::set_revert_account_info`, that takes `Option<Option<AccountInfo>>`, is that the one we should be calling here?

Comment (emhane) [reply crates/storage/provider/src/providers/database/provider.rs:671]: @DaniPopes, you're probably familiar with the revm api and can answer this too?

Comment (DaniPopes) [reply crates/storage/provider/src/providers/database/provider.rs:671]: I'm not familiar with state/transitions/bundles APIs

**Comment by rkrasiuk** [comment]
as i mentioned in reviews for #8264, bundle builder API must be improved before integration in reth

**Pull request closed without merging**